lsp-mode.el

```diff
@@ -6280,8 +6280,8 @@ WORKSPACE is the active workspace."
                                      (/ (nth 2 (time-since before-send)) 1000))
                 workspace))
              (when callback
-               (funcall callback (lsp:json-response-result json-data))
-               (remhash id (lsp--client-response-handlers client)))))
+               (remhash id (lsp--client-response-handlers client))
+               (funcall callback (lsp:json-response-result json-data)))))
```

Comment:
The issue is in lsp-request-while-no-input - it throws and in that way, we skip the cleanup portion.

Reply:
Having to throw inside callback is indeed a disaster. Especially when the callback's throw happens, it will also prevent other messages from being processed in https://github.com/emacs-lsp/lsp-mode/blob/master/lsp-mode.el#L6367-L6369
The throw is introduced by #2810, maybe we should revert that and introduce another mechanism to notify `sit-for` or `accept-process-output`

Reply:
If the callback is processed in a timer, can we still capture the throw inside the callback?

Reply:
yes, because the main loop will be in the `sit-for`

Reply:
Ok, I can confirm that if we have more than one message and we use while-no-input method we will lose the second message. I propose merging this because it will make things better and it will remove the leak. The message lossage can be addressed separately.

Reply:
ok